*** a/mariadb-5.5.23/scripts/mysqld_safe.sh	2012-05-02 17:01:03.521496156 -0700
--- b/mariadb-5.5.23/scripts/mysqld_safe.sh	2012-05-02 16:58:23.771495589 -0700
***************
*** 18,23 ****
--- 18,25 ----
  nowatch=0
  mysqld_ld_preload=
  mysqld_ld_library_path=
+ flush_caches=0
+ numa_interleave=0
  
  # Initial logging status: error log is not open, and not using syslog
  logging=init
***************
*** 63,68 ****
--- 65,73 ----
    --syslog                   Log messages to syslog with 'logger'
    --skip-syslog              Log messages to error log (default)
    --syslog-tag=TAG           Pass -t "mysqld-TAG" to 'logger'
+   --flush-caches             Flush and purge buffers/caches
+   --numa-interleave          Run mysqld with its memory interleaved
+                              on all CPUs
  
  All other options are passed to the mysqld program.
  
***************
*** 221,226 ****
--- 226,233 ----
        --skip-syslog) want_syslog=0 ;;
        --syslog-tag=*) syslog_tag="$val" ;;
        --timezone=*) TZ="$val"; export TZ; ;;
+       --flush-caches) flush_caches=1 ;;
+       --numa-interleave) numa_interleave=1 ;;
  
        --help) usage ;;
  
***************
*** 731,736 ****
--- 738,778 ----
  fi
  
  #
+ # Flush and purge buffers/caches.
+ #
+ 
+ if @TARGET_LINUX@ && test $flush_caches -eq 1
+ then
+   # Locate sync, ensure it exists.
+   if ! my_which sync > /dev/null 2>&1
+   then
+     log_error "sync command not found, required for --flush-caches"
+     exit 1
+   # Flush file system buffers.
+   elif ! sync
+   then
+     # Huh, the sync() function is always successful...
+     log_error "sync failed, check if sync is properly installed"
+   fi
+ 
+   # Locate sysctl, ensure it exists.
+   if ! my_which sysctl > /dev/null 2>&1
+   then
+     log_error "sysctl command not found, required for --flush-caches"
+     exit 1
+   # Purge page cache, dentries and inodes.
+   elif ! sysctl -q -w vm.drop_caches=3
+   then
+     log_error "sysctl failed, check the error message for details"
+     exit 1
+   fi
+ elif test $flush_caches -eq 1
+ then
+   log_error "--flush-caches is not supported on this platform"
+   exit 1
+ fi
+ 
+ #
  # Uncomment the following lines if you want all tables to be automatically
  # checked and repaired during startup. You should add sensible key_buffer
  # and sort_buffer values to my.cnf to improve check performance or require
***************
*** 750,755 ****
--- 792,822 ----
  
  cmd="`mysqld_ld_preload_text`$NOHUP_NICENESS"
  
+ #
+ # Set mysqld's memory interleave policy.
+ #
+ 
+ if @TARGET_LINUX@ && test $numa_interleave -eq 1
+ then
+   # Locate numactl, ensure it exists.
+   if ! my_which numactl > /dev/null 2>&1
+   then
+     log_error "numactl command not found, required for --numa-interleave"
+     exit 1
+   # Attempt to run a command, ensure it works.
+   elif ! numactl --interleave=all true
+   then
+     log_error "numactl failed, check if numactl is properly installed"
+   fi
+ 
+   # Launch mysqld with numactl.
+   cmd="$cmd numactl --interleave=all"
+ elif test $numa_interleave -eq 1
+ then
+   log_error "--numa-interleave is not supported on this platform"
+   exit 1
+ fi
+ 
  for i in  "$ledir/$MYSQLD" "$defaults" "--basedir=$MY_BASEDIR_VERSION" \
    "--datadir=$DATADIR" "--plugin-dir=$plugin_dir" "$USER_OPTION"
  do
